server/plugin/hooks.go

```diff
@@ -0,0 +1,80 @@
+package plugin
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/mattermost/mattermost-server/v5/model"
+	"github.com/mattermost/mattermost-server/v5/plugin"
+)
+
+type FileStruct struct {
+	ID     string
+	Expiry time.Time
+}
+
+func (p *Plugin) MessageHasBeenPosted(c *plugin.Context, post *model.Post) {
+	// If the message is posted by bot simply return
+	if post.UserId == p.botUserID {
+		return
+	}
+
+	channel, appErr := p.API.GetChannel(post.ChannelId)
```

Comment:
I am concerned about the performance overhead of executing an RPC (GetChannel) on practically every message posted. @hanzei it appears that all that is needed here is to determine if the channel is direct; can it be done from the Channel ID itself? Is there a convention?

Reply:
I don't think there is a specific convention for direct channel IDs. They use the same 25 random character ids. The names, however, are special. But that is a piece of information that is not available right away without an RPC call.

Reply:
I agree that an RPC call for every message posted is a performance concern. The only solution that comes to my mind is a cache.

Reply:
What kind of cache are you thinking about using here? There are two options:
1. Using the KV store to store the DM channel ids for each user. This will be slower.
2. Using an in-memory cache, but I guess this will need separate handling for multiple clusters.

Reply:
1/5 In this case a simple, (limited) in-memory cache that warms up over time may be sufficient; separate on each node. The information in question (is a DM channel) never changes for a given ID. Just a sync.Map(string->bool)

Reply:
:+1: on @levb's suggestion. One technical challenge is to design the cache in a way that does not consume too much memory. A TTL for each entry could help.

Reply:
Sure @levb. We'll implement this and create a PR as soon as possible.

Reply:
@hanzei @levb Can we use packages like https://github.com/allegro/bigcache?

Reply:
https://github.com/allegro/bigcache doesn't quite seem like the right package to use, as its focus is on big entries (e.g. byte slices are used in the API), while in our case we only want to store booleans.

Reply:
Or should we implement it ourselves?

Reply:
https://github.com/bluele/gcache look fine to me. Implementing one is also an option. In the end, you pick what you think fits the use case and what you are comfortable working with.

Reply:
I think that all that is needed is a `sync.Map` populated as map[string]bool (channelID->isDM), lazily populated, and checked first. 1/5 it's not worth bringing in a 3rd party package.

Reply:
BTW, I implemented this cache with the package above. So, should I revert that change and use a `sync.Map` instead?

Reply:
You can also check the PR here - https://github.com/Brightscout/mattermost-plugin-servicenow-virtual-agent/pull/3